plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -287,14 +300,45 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
         variableAssignment: FirVariableAssignment,
         data: StmtConversionContext<ResultTrackingContext>,
     ): Exp {
+
         // It is not entirely clear whether we can get away with ignoring the distinction between
         // lvalues and rvalues, but let's try to at first, and we'll fix it later if it turns out
         // not to work.
-        val convertedLValue = data.convert(variableAssignment.lValue)
-        val lValueType = data.embedType(variableAssignment.lValue)
-        val convertedRValue = data.convert(variableAssignment.rValue)
-        val rValueType = data.embedType(variableAssignment.rValue)
-        data.addStatement(Stmt.assign(convertedLValue, convertedRValue.convertType(rValueType, lValueType)))
+
+        if (variableAssignment.isAssigningToClassField) {
+
+            val lValue = variableAssignment.lValue as FirPropertyAccessExpression
+            val lValueSymbol = lValue.calleeSymbol as FirPropertySymbol
+            val lValueType = data.embedType(variableAssignment.lValue)
+            val receiver = data.convert(lValue.dispatchReceiver)
+
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
+
+            // According to the setter symbol, we should invoke a method, or assigning a value
+            val setterSymbol = lValueSymbol.setterSymbol!!
+            if (setterSymbol.hasBody) {
```

Comment:
Why do we decide what to do based on whether the setter has a body here, but use the existence of a backing field elsewhere?

Reply:
Is it even possible that this occurs? I would imagine that if there was no body then `lValueSymbol.setterSymbol` is just `null`. Or does that indicate something else?

Reply:
@22slin22 this is needed to distinguish the cases like:
```kotlin
class Foo(var b: Int) {
    var a: Int = 0
       set(value) { field = value + 1}
}
```

In this example `b` does not have a body, but when we set it we should use a field assignment:
```viper
inhale acc(f.b)
f.b = 1
exhale acc(f.b)
```

Instead, when the setter has a body, we have to invoke the method corresponding to it:

```viper
anon = Foo_a_setter(n)
```

Reply:
I think for this (and also the corresponding getter code), Ilya's opinion on how to get the correct behaviour would be valuable.